examples/pytorch/dagnn/main.py

```diff
@@ -0,0 +1,272 @@
+
+            for _ in range(self.k):
+                feats = feats * norm
+                graph.ndata['h'] = feats
+                graph.update_all(fn.copy_u('h', 'm'),
+                             fn.sum('m', 'h'))
+                feats = graph.ndata.pop('h')
+                feats = feats * norm
+                results.append(feats)
+
+            H = torch.stack(results, dim=1)
+            S = F.sigmoid(torch.matmul(H, self.s))
+            S = S.permute(0, 2, 1)
+            H = torch.matmul(S, H).squeeze()
+
+            return H
+
+
+class MLPLayer(nn.Module):
+    def __init__(self,
+                 in_dim,
+                 out_dim,
+                 bias=True,
+                 activation=None,
+                 batchnorm=False,
+                 dropout=0,
+                 drop_bef=True):
+        super(MLPLayer, self).__init__()
+
+        self.linear = nn.Linear(in_dim, out_dim, bias=bias)
+        self.activation = activation
+        self.batchnorm = batchnorm
+        if batchnorm:
+            self.bn = nn.BatchNorm1d(out_dim)
+        self.dropout = nn.Dropout(dropout)
+        self.drop_bef = drop_bef
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        gain = 1.
+        if self.activation is F.relu:
+            gain = nn.init.calculate_gain('relu')
+        nn.init.xavier_uniform_(self.linear.weight, gain=gain)
+        if self.linear.bias is not None:
+            nn.init.zeros_(self.linear.bias)
+        if self.batchnorm:
+            self.bn.reset_parameters()
+
+    def forward(self, feats):
+
+        if self.drop_bef:
+            feats = self.dropout(feats)
+            feats = self.linear(feats)
+            if self.batchnorm:
+                feats = self.bn(feats)
+            if self.activation:
+                feats = self.activation(feats)
+        else:
+            feats = self.linear(feats)
+            if self.batchnorm:
+                feats = self.bn(feats)
+            if self.activation:
+                feats = self.activation(feats)
+            feats = self.dropout(feats)
+
+        return feats
+
+
+class DAGNN(nn.Module):
+    def __init__(self,
+                 k,
+                 in_dim,
+                 hid_dim,
+                 out_dim,
+                 bias=True,
+                 activation=F.relu,
+                 batchnorm=False,
+                 dropout=0,
+                 drop_bef=True):
+        super(DAGNN, self).__init__()
+        self.mlp = nn.ModuleList()
+        self.mlp.append(MLPLayer(in_dim=in_dim, out_dim=hid_dim, bias=bias,
+                                 activation=activation, batchnorm=batchnorm,
+                                 dropout=dropout, drop_bef=drop_bef))
+        self.mlp.append(MLPLayer(in_dim=hid_dim, out_dim=out_dim, bias=bias,
+                                 activation=None, batchnorm=batchnorm,
+                                 dropout=dropout, drop_bef=drop_bef))
```

Comment:
I guess you need to disable dropout for the last layer.

Reply:
Here dropout locates before transform features.